### ISSUE TITLE
Making MMEX categories more useful: automatic assignment & color coding

### ISSUE THREAD

**Issue by jensb** [moneymanagerex/moneymanagerex#4268]
Copied from forum post (which was then deleted, to track all communication related to this issue here)

I just got started with MMEX and I like the approach. I was using iFinance (Mac) before but now my primary computer is Linux, and iFinance also seems not too well maintained (no update in several years). So I'm planning to switch.

But before that, MMEX will need to get killer features: automatic category assignment based on keywords, and automatic coloring of transactions based on category. For starters, I'll plan to implement those two. I'm asking for some advice how to get started.

To me, this looks like a four-step task:

1. Database migration: CATEGORY_V1 will need three more string columns: BGCOLOR, FGCOLOR, KEYWORDS.
How do I add these and ensure MMEX will pick up the new format upon startup? Just add another .sql file to the database/ folder?

2. Import & Create hook: When importing or creating transactions, KEYWORDS will need to be scanned and if recipient or notes or any string custom fields match any keyword, the category_id needs to be set. When editing transactions, this should not happen to enable users to still override the automatic assignment.
What would be the best practice to insert this code? Preferably (of couse) in one central place, as a before-save or before-commit hook. Also, selecting a number of transactions and replaying automatic category assignment (e.g. after adding a keyword) should be possible using a context menu item.

3. Display. When displaying transactions, the colors need to be applied to the transactions grid. Where would I insert this?

4. Tests. The test suite needs to be updated to ensure that keywords are correctly read, edits cause category rescans, no race conditions occur when editing AND manually setting the category, and so on. I found a Travis CI config, but no test suite, where is it?

I would appreciate any hints where to start looking.
I have some ideas but I'd like to avoid going into a completely absurd direction by mistake.

**Comment by whalley** [comment]
1. DB guidance: https://github.com/moneymanagerex/moneymanagerex/blob/master/src/db/readme.txt

2. Import/Export code is here: https://github.com/moneymanagerex/moneymanagerex/tree/master/src/import_export

3. Need to consider how 'split transactions' are colored. The main transactions grid is generated here - https://github.com/moneymanagerex/moneymanagerex/blob/master/src/mmchecking_list.cpp. Need to also consider what you want to do with asset and stocks/shares transactions. Also, a recurring transaction list needs to be considered. Do you also want to color budgets for various categories in the budget setup?

4. There is no automated test suite as far as I'm aware!

You also need to consider:

5. Maintenance of the colors and keywords in the category dialog: https://github.com/moneymanagerex/moneymanagerex/blob/master/src/categdialog.cpp

6. If you want to use the colors for the reports.

7. How the coloring interacts with the existing transaction coloring that is supported by the application.  Unfortunately, I don't think the ListView in wxWidgets supports individual cell color highlighting so we couldn't highlight just the category with the color.

I would suggest further refining the requirements/behavior and generating a few mockups of the changes. Sure there are a few other items to consider. 

If you are going to develop this then I would suggest splitting the keyword matching and changes to import logic from the category coloring as they are two independent features.

**Comment by vomikan** [comment]
https://github.com/moneymanagerex/moneymanagerex/issues/4216

**Comment by jensb** [comment]
1. Thanks
2. Looked at it, can parse most of it, but broke everything trying to do small modifications. Especially dialog & UI design is weird. My C++ is limited unfortunately, I'm more at home with web development and scripting languages. If MMEX were written in Javascript I'd probably have it half finished by now. :-)
3. Good point. Here's an example mockup which works if we just color the category column. This is also good because then the "selected row" bgcolor can be applied to all remaining columns and there's confusion as to what is selected.
![image](https://user-images.githubusercontent.com/44540/160702525-a7ae9742-2f24-4e9d-b3cb-4f77240d4bda.png)
4. Ouch. That should definitely be changed, especially for a finance app ...

5. Mockup:
![image](https://user-images.githubusercontent.com/44540/160706192-c8ec96a6-bb89-4d50-8c2d-f738e5f46406.png)
Actually, I'd prefer to have one window for browsing and editing with automatic commitment of changes without "OK" button:
![image](https://user-images.githubusercontent.com/44540/160706936-ebb61a63-dab5-4236-96c2-ae3333359226.png)
Initially, the right side elements would be empty & disabled. On click on a category, or the "+" icon, the right side would be activated. "Close" button is not required since the window controls can do this already. Checking one of the "Inherit" checkboxes will disable the respective color selector.
6. Not yet. Maybe in a later revision. They should be available as properties for template authors though.
7. My thought was that category colors only color the category column. This would avoid conflicts and users would be free to choose what they want to use. If the UI doesn't support single cell coloring, that would be too bad.

PS: The mockups were done using moqups.com. I can share a (private) link for collaboration on these mockups if you like. I don't work there or anything, I just use their service.

**Comment by jensb** [comment]
I just noticed you can assign payees to categories as well.
Is this meant to achieve a similar thing, that a payee will automatically lead to assignment of a category if it appears again in a future transaction?
This would be partly duplication of functionailty and should be disabled when using the above approach (which is much more flexible).

**Comment by whalley** [comment]
> 7\. My thought was that category colors only color the category column. This would avoid conflicts and users would be free to choose what they want to use. If the UI doesn't support single cell coloring, that would be too bad.

There are other controls such as wxGrid that I think will support this flexibility but it would require further investigation.

**Comment by whalley** [comment]
> I just noticed you can assign payees to categories as well. Is this meant to achieve a similar thing, that a payee will automatically lead to assignment of a category if it appears again in a future transaction? This would be partly duplication of functionailty and should be disabled when using the above approach (which is much more flexible).

Yes, you can assign a default category to a payee or specify if the 'last used for that payee' is retained.

**Comment by jensb** [comment]
> > 7. My thought was that category colors only color the category column. This would avoid conflicts and users would be free to choose what they want to use. If the UI doesn't support single cell coloring, that would be too bad.
> 
> There are other controls such as wxGrid that I think will support this flexibility but it would require further investigation. One other option may be to dynamically create a colored cell/image and add that, perhaps as a category color column, so that those interested in category colors they can include in the view.

I like this approach, it's compact and we can even define colored icons / avatars / whatever instead of having to color the cell itself. For example ■ (&#9632;) or even font-awesome icons (https://fontawesome.com/search?m=free&s=solid). That would probably avoid having to change the grid concept.

> Yes, you can assign a default category to a payee or specify if the 'last used for that payee' is retained.

I didn't even notice, and this is good, but I think we can do better. :-)

For example, I have one insurance company (payee) with three different contracts which I want to keep separated by using categories (which are then used in reports to watch my payments and reimbursements). This is only possible by looking at the notes column too, but since this column *also* always contains the current month, it must be a substring match. And vice versa, we regularly buy in a supermarket which always has a slightly different payee name (Toom, Toom sagt Danke, Toom wünscht Frohe Ostern, Toom $RANDOM_NUMBER, Toom jetzt neu, Toom whatever) but this is all the same "Food" category.

But it is possible to seamlessly upgrade default categories for payees to the scheme mentioned above by just adding a keyword to each category that is exactly equal to the payee name. Categorization would just keep working for existing users. I would be very interested in such a change and I could surely take care of writing the SQL data migration code, but I couldn't do the C++ part by myself.

What do you think?

**Comment by jensb** [comment]
Any thoughts on this?
